drivers/gpio/gpio_pcal6416a.c

```diff
@@ -0,0 +1,269 @@
+	PCAL6416A_REG_CONFIGURATION_0 = 0x06,
+	PCAL6416A_REG_CONFIGURATION_1 = 0x07,
+	PCAL6416A_REG_OUTPUT_DRIVE_STRENGTH_0_0 = 0x40,
+	PCAL6416A_REG_OUTPUT_DRIVE_STRENGTH_0_1 = 0x41,
+	PCAL6416A_REG_OUTPUT_DRIVE_STRENGTH_1_0 = 0x42,
+	PCAL6416A_REG_OUTPUT_DRIVE_STRENGTH_1_1 = 0x43,
+	PCAL6416A_REG_INPUT_LATCH_0 = 0x44,
+	PCAL6416A_REG_INPUT_LATCH_1 = 0x45,
+	PCAL6416A_REG_PULL_UP_DOWN_ENABLE_0 = 0x46,
+	PCAL6416A_REG_PULL_UP_DOWN_ENABLE_1 = 0x47,
+	PCAL6416A_REG_PULL_UP_DOWN_SELECT_0 = 0x48,
+	PCAL6416A_REG_PULL_UP_DOWN_SELECT_1 = 0x49,
+	PCAL6416A_REG_INTERRUPT_MASK_0 = 0x4A,
+	PCAL6416A_REG_INTERRUPT_MASK_1 = 0x4B,
+	PCAL6416A_REG_INTERRUPT_STATUS_0 = 0x4C,
+	PCAL6416A_REG_INTERRUPT_STATUS_1 = 0x4D,
+	PCAL6416A_REG_OUTPUT_PORT_CONFIGURATION = 0x4F,
+};
+
+static int pcal6416a_i2c_write(const struct i2c_dt_spec *i2c, uint8_t register_address,
+			       uint8_t value)
+{
+	int rc;
+
+	LOG_DBG("writing to register 0x%02X value 0x%02X", register_address, value);
+	rc = i2c_reg_write_byte_dt(i2c, register_address, value);
+
+	if (rc != 0) {
+		LOG_ERR("unable to write to register 0x%02X, error %i", register_address, rc);
+	}
+
+	return rc;
+}
+
+static int pcal6416a_i2c_read(const struct i2c_dt_spec *i2c, uint8_t register_address,
+			      uint8_t *value)
+{
+	int rc;
+
+	rc = i2c_reg_read_byte_dt(i2c, register_address, value);
+	LOG_DBG("reading from register 0x%02X value 0x%02X", register_address, *value);
+
+	if (rc != 0) {
+		LOG_ERR("unable to read from register 0x%02X, error %i", register_address, rc);
+	}
+
+	return rc;
+}
+
+static int pcal6416a_pins_cfg_apply(const struct i2c_dt_spec *i2c,
+				    const struct pcal64xxa_pins_cfg *pins_cfg)
+{
+	int rc;
+
+	rc = pcal6416a_i2c_write(i2c, PCAL6416A_REG_PULL_UP_DOWN_SELECT_0,
+				 (uint8_t)pins_cfg->pull_ups_selected);
+	if (rc != 0) {
+		return -EIO;
+	}
+
+	rc = pcal6416a_i2c_write(i2c, PCAL6416A_REG_PULL_UP_DOWN_SELECT_1,
+				 (uint8_t)(pins_cfg->pull_ups_selected >> 8));
+	if (rc != 0) {
+		return -EIO;
+	}
+
+	rc = pcal6416a_i2c_write(i2c, PCAL6416A_REG_PULL_UP_DOWN_ENABLE_0,
+				 (uint8_t)pins_cfg->pulls_enabled);
+	if (rc != 0) {
+		return -EIO;
+	}
+
+	rc = pcal6416a_i2c_write(i2c, PCAL6416A_REG_PULL_UP_DOWN_ENABLE_1,
+				 (uint8_t)(pins_cfg->pulls_enabled >> 8));
+	if (rc != 0) {
+		return -EIO;
+	}
+
+	rc = pcal6416a_i2c_write(i2c, PCAL6416A_REG_OUTPUT_PORT_0, (uint8_t)pins_cfg->outputs_high);
+	if (rc != 0) {
+		return -EIO;
+	}
+
+	rc = pcal6416a_i2c_write(i2c, PCAL6416A_REG_OUTPUT_PORT_1,
+				 (uint8_t)(pins_cfg->outputs_high >> 8));
+	if (rc != 0) {
+		return -EIO;
+	}
+
+	rc = pcal6416a_i2c_write(i2c, PCAL6416A_REG_CONFIGURATION_0,
+				 (uint8_t)pins_cfg->configured_as_inputs);
+	if (rc != 0) {
+		return -EIO;
+	}
+
+	rc = pcal6416a_i2c_write(i2c, PCAL6416A_REG_CONFIGURATION_1,
+				 (uint8_t)(pins_cfg->configured_as_inputs >> 8));
+	if (rc != 0) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int pcal6416a_inputs_read(const struct i2c_dt_spec *i2c, uint16_t *int_sources,
```

Comment:
Please add the complementary `pcal6408a_inputs_read`

Reply:
done